src/database.c

```diff
@@ -231,18 +238,20 @@ static int write_db(const char *index, const char *data)
 		return 2;
 	}
 
-    len = strlen(index);
-    if (len > MAX_PATH_LEN) { /* key size is greater than LMDB key limit */
-        hash = path_to_hash(index);
-        key.mv_data = (void *)hash;
-        key.mv_size = gcry_md_get_algo_dlen(GCRY_MD_SHA256) * 2 + 1;
-    }
-    else {
-        key.mv_data = (void *)index;
-	    key.mv_size = len;
-    }
-    value.mv_data = (void *)data;
+	len = strlen(index);
+	if (len > MDB_MAXKEYSIZE) { /* key size is greater than LMDB key limit */
+		hash = path_to_hash(index, len);
+		if (hash == NULL)
+			return 5;
+		key.mv_data = (void *)hash;
+		key.mv_size = gcry_md_get_algo_dlen(GCRY_MD_SHA512) * 2 + 1;
+	} else {
+		key.mv_data = (void *)index;
+		key.mv_size = len;
+	}
+	value.mv_data = (void *)data;
```

Comment:
This branch and the below look exactly the same. Generally, I'd factor out any common code into a helper in order to reduce code duplication wherever possible. Maybe you could do this too, but it's up to you.

Reply:
I'd prefer to see the refactoring as a separate pull request. Let's get this hashing in place and then make it more efficient with a followup patch.

Reply:
Yeah, could do that too.